### PR TITLE
bpo-33668: Fix inspect.getmembers to traverse mro when object is class

### PR DESCRIPTION
**Pull request by corona10**
https://bugs.python.org/issue33668

### CODE REVIEW

**Comment by corona10** [comment]
@serhiy-storchaka 
To handle this case,

``` python
import inspect
import types

class Base(object):
    @types.DynamicClassAttribute
    def fried(self):
        return self._x
    _x = object()

class A(Base):
    @types.DynamicClassAttribute
    def eggs(self):
        return self._x
    _x = object()

inspect.getmembers(A)
```

**Comment by corona10** [comment]
For reviewers,

> __bases__ is a tuple containing the base classes from https://docs.python.org/3/reference/datamodel.html

I add the check the type of object.__bases__.

**Comment by corona10** [comment]
@serhiy-storchaka
But I don't know exactly why this is needed. 
The first reference is here. https://github.com/python/cpython/commit/e03ea37a7bea48c46e6d96851f471db0f3c8e6e2#diff-8fb5d890c8ae80e19883adc7fbdd2c8fR337

**Comment by corona10** [comment]
If we run this script

``` python
import inspect
import types

class Base(object):
    @types.DynamicClassAttribute
    def fried(self):
        return self._x
    _x = object()

class A(Base):
    @types.DynamicClassAttribute
    def eggs(self):
        return self._x
    _x = object()

inspect.getmembers(A)
```

We get duplicated members,
```
('fried', <types.DynamicClassAttribute object at 0x100747970>), ('fried', <types.DynamicClassAttribute object at 0x100747970>)]
```

IMHO, we should add a code to check this is duplicated.
```
 329     try:
 330         if isinstance(object.__bases__, tuple):
 331             for base in object.__bases__:
 332                 for k, v in base.__dict__.items():
 333                     if isinstance(v, types.DynamicClassAttribute):
 334                          if k not in names:
 335                              names.append(k)
 336     except AttributeError:
```

**Comment by corona10** [comment]
I found the case of the related to this code.
``` bash
"Lib/test/test_enum.py", line 2560, in test_inspect_getmembers
```

I think that we can update this code into
```
@@ -331,7 +331,8 @@ def getmembers(object, predicate=None):
             for base in object.__bases__:
                 for k, v in base.__dict__.items():
                     if isinstance(v, types.DynamicClassAttribute):
-                        names.append(k)
+                        if k not in names:
+                            names.append(k)
     except AttributeError:
```

**Comment by corona10** [comment]
@serhiy-storchaka 

Current implementation returns dulplicated members,
```
, ('eggs', <types.DynamicClassAttribute object at 0x10ed38970>), ('fried', <types.DynamicClassAttribute object at 0x10ed38940>), ('fried', <types.DynamicClassAttribute object at 0x10ed38940>)]
```

But with this PR
```
('_x', <object object at 0x10e1691a0>), ('eggs', <types.DynamicClassAttribute object at 0x10e2c0970>), ('fried', <types.DynamicClassAttribute object at 0x10e2c0940>)]
```

It works well.

**Comment by serhiy-storchaka** [comment]
Shouldn't this be executed only for classes?

**Comment by corona10** [comment]
@serhiy-storchaka 
``` python
import enum
import inspect

class Color(enum.Enum):
    red = 1
    green = 2
    blue = 3

values = dict((
    ('__class__', enum.EnumMeta),
    ('__doc__', 'An enumeration.'),
    ('__members__', Color.__members__),
    ('__module__', __name__),
    ('blue', Color.blue),
    ('green', Color.green),
    ('name', enum.Enum.__dict__['name']),
    ('red', Color.red),
    ('value', enum.Enum.__dict__['value']),
    ))

print(dict(inspect.getmembers(Color)).keys())
c = Color.red
print(c.value)
```

Here is the case,
Without traversing __base__, it can not get `value` from getmembers.

**Comment by serhiy-storchaka** [comment]
In all your examples `object` is a class.

**Comment by corona10** [comment]
@serhiy-storchaka 
I am now understanding!
I've updated to check the object is class before traversing.

File: Lib/inspect.py
Comment (serhiy-storchaka) [inline Lib/inspect.py]: I suggest to move the above comment inside the `if` branch. Seems `:dd` is a typo of `Add`.

This `if isclass(object):` can be merged with other `if isclass(object):` at the beginning of the function.

File: Lib/inspect.py
Comment (serhiy-storchaka) [inline Lib/inspect.py]: Is this still needed?

**Comment by serhiy-storchaka** [comment]
> Current implementation returns dulplicated members,

The comment above the loop over `object.__bases__` explicitly says about duplicated entries. And the code below handles them.

**Comment by corona10** [comment]
@serhiy-storchaka Thanks I've updated.

File: Misc/NEWS.d/next/Library/2018-07-14-23-52-12.bpo-33668.7poHxA.rst
Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/Library/2018-07-14-23-52-12.bpo-33668.7poHxA.rst]: `travering` -> `traversing`

And I would add "only" before "traversing".

And maybe add a markup: ``` :func:`inspect.getmembers` ```.

File: Lib/test/test_inspect.py
Comment (serhiy-storchaka) [inline Lib/test/test_inspect.py]: I think it would be better to change the name (and perhaps the code) of the test. There is nothing specific for modules. Instead we need to test that the instance attribute `__bases__` doesn't break `inspect.getmembers()`.

**Comment by corona10** [comment]
@serhiy-storchaka 
Thanks, I've updated the documentation and the unit test.

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by corona10** [comment]
@1st1 

Can you take a look?

File: Lib/inspect.py
Comment (1st1) [inline Lib/inspect.py]: Line longer than 79 chars; please fix the PR to follow PEP 8

**Comment by taleinat** [comment]
@corona10, in this case I think it would be prudent to wait for @1st1 to review this, given that his review has been requested both by @serhiy-storchaka here and by @ilevkivskyi in a comment on the b.p.o. issue.

Especially considering that many people have been on lengthy vacations recently, I find it appropriate to wait some more.

Feel free to ping me again in a month if this continues to linger.

File: Lib/test/test_inspect.py
Comment (1st1) [inline Lib/test/test_inspect.py]: The `types` module is already imported in `inspect.py`. Why another import?

File: Lib/inspect.py
Comment (1st1) [inline Lib/inspect.py]: There's no need to prepend the class to its MRO here: `(object,) + getmro(object)`; the class is always part of its own MRO:

Try running:
```python
class A: pass
print(A.__mro__)
```

I suggest to try to fix this and see how that works out.

File: Lib/inspect.py
Comment (1st1) [reply Lib/inspect.py]: BTW, I remember leaving this kind of comment on almost all of your PRs. Please configure your editor to trim the whitespace and show you a ruler at 79 characters.

File: Lib/test/test_inspect.py
Comment (1st1) [inline Lib/test/test_inspect.py:1182]: I'd split this test in two (they are testing different behaviours)

**Review by 1st1 — CHANGES_REQUESTED** [review]
@corona10 See the review comments.

@serhiy-storchaka Are you sure you want to backport this to 2.7?  I'm not; it might accidentally fix some things that weren't working for a decade and there are now workarounds in place that might fail if we fix this.

**Comment by serhiy-storchaka** [comment]
@1st1 No, I am not sure.

File: Lib/inspect.py
Comment (corona10) [reply Lib/inspect.py]: @1st1 
I'm sorry, it always seems like I made a mistake.

**Comment by corona10** [comment]
@1st1 

Thank you for your review despite the busy situation.
Following your guide, I 've added a configuration to my .vimrc also! Thanks!

``` .vimrc
 set colorcolumn=79
 autocmd BufWritePre *.py :%s/\s\+$//e
```

PTAL

**Comment by taleinat** [comment]
@1st1, @serhiy-storchaka

> @serhiy-storchaka Are you sure you want to backport this to 2.7? I'm not; it might accidentally fix some things that weren't working for a decade and there are now workarounds in place that might fail if we fix this.

Isn't this true for 3.6 and 3.7 as well, except for the "decade" part?

IMO this should only go into 3.8, and have a "What's New" entry mentioning the change.

This strikes me as being similar to bpo-33899, in that this is probably how it should have been to begin with, but changing it now in released versions would likely unexpectedly break existing code that uses it.

File: Lib/inspect.py
Comment (serhiy-storchaka) [inline Lib/inspect.py:329]: Should not it iterate `mro` instead of `object.__bases__`? Or `_static_getmro(object)`?

**Comment by serhiy-storchaka** [comment]
This PR should be backported at least to 3.7, because a module-level `__getattr__` is new in 3.7, and this increase the chance to encounter this bug.

But the bug can be encountered in older Python versions for classes with `__getattr__`. I don't think that fixing it will break somebody's code. Do you have any examples?

**Comment by taleinat** [comment]
> Do you have any examples?

No, I don't. You're probably right about backporting this to 3.x.

File: Lib/inspect.py
Comment (taleinat) [reply Lib/inspect.py:329]: Indeed, it should apparently use one of those.

**Comment by corona10** [comment]
@1st1
Can you take a look, please?

**Comment by serhiy-storchaka** [comment]
Could you please add a test for your last change? I.e. the test that passed when iterate mro, but fails when iterate `__bases__`. It can be a class with indirectly inherited DynamicClassAttribute.
```python
class A: # A DynamicClassAttribute should be defined here
class B(A): pass
class C(B): pass
# test inspect.getmembers(C)
```

**Comment by corona10** [comment]
@serhiy-storchaka 
Thanks
I 've added a test `test_getmembers_with_dynamic_class_attribute` which was failed when
iterate `__bases__`.

File: Lib/test/test_inspect.py
Comment (serhiy-storchaka) [inline Lib/test/test_inspect.py]: Why not `self.assertIn('ham', attrs)`?

Add also test for `inspect.getmembers(B)`.

Comment (corona10) [reply Lib/test/test_inspect.py]: @serhiy-storchaka 
Just checking `self.assertIn('ham', attrs)` does not fail when iterate `__bases__`.

**Comment by corona10** [comment]
@serhiy-storchaka 

I updated the unit test and to understand more deeply.
Please reference below script and outputs.
As I am not an expert on inspect module. I don't know which is right output.

```` python
import types
import inspect


class A:
    @types.DynamicClassAttribute
    def ham(self):
        return 'eggs'

class B(A): pass
class C(B): pass

print([a for a in inspect.getmembers(A) if a[0] == 'ham'])
print([b for b in inspect.getmembers(B) if b[0] == 'ham'])
print([c for c in inspect.getmembers(C) if c[0] == 'ham'])
````

### AS-IS
````
bash
[('ham', <types.DynamicClassAttribute object at 0x10d2f8198>)]
[('ham', <types.DynamicClassAttribute object at 0x10d2f8198>), ('ham', <types.DynamicClassAttribute object at 0x10d2f8198>)]
[('ham', <types.DynamicClassAttribute object at 0x10d2f8198>)]
````

### TOBE 98d79c7
````
[('ham', <types.DynamicClassAttribute object at 0x103365130>), ('ham', <types.DynamicClassAttribute object at 0x103365130>)]
[('ham', <types.DynamicClassAttribute object at 0x103365130>), ('ham', <types.DynamicClassAttribute object at 0x103365130>)]
[('ham', <types.DynamicClassAttribute object at 0x103365130>), ('ham', <types.DynamicClassAttribute object at 0x103365130>)]
````

**Comment by serhiy-storchaka** [comment]
Should not `inspect.getmembers()` remove duplicates?

Seems using mro instead of `__bases__` doesn't fix any bug. In this case it is better to restore the original code.

**Comment by corona10** [comment]
@serhiy-storchaka 
I revert the code to use `__bases__`.
But I was wondering if this PR was the right way. 
Do you think this PR would not be merged and should we find out another way?
If so, I am going to close this PR :)

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by vstinner** [comment]
I removed the " needs backport to 3.6" label, the 3.6 branch no longer accept bugfixes (only security fixes are accepted): https://devguide.python.org/#status-of-python-branches

**Comment by corona10** [comment]
@1st1 Can you take a look, please?

**Comment by corona10** [comment]
@1st1 @serhiy-storchaka @vstinner @csabella

Hi, this PR remains for a year.
Is there any plan with this PR?

Thanks

**Comment by corona10** [comment]
@1st1 @serhiy-storchaka @vstinner @csabella

I close this PR due to no feedback..
If there is any chance to work with this issue.
I will reopen this PR

Thanks!

**Pull request closed without merging**
